### ISSUE TITLE
Move away from AWS IPv6 prefixes for CML external connectivity

### ISSUE THREAD

**Issue by cmm-cisco** [becomingahacker/cloud-cml#1]
DHCPv6-PD is broken at this time on AWS.

* [Note](https://github.com/cmm-cisco/cloud-cml/commit/027a8026c686e0e29b492943af6c4334e637c5c3)

As @eptiger notes in [this issue](https://github.com/aws/karpenter-provider-aws/issues/2187#issuecomment-1201961777
), it's not currently possible to specify an `ipv6-address-count` and `ipv6-prefix-count` at the same time.  Adding an IPv6 prefix used to also add a primary IPv6 address when using [Terraform `aws_network_interface` resources](https://github.com/cmm-cisco/cloud-cml/blob/166ae43b9f91fdc535932da109031c3ff1a2f28b/module-cml2-ec2-instance/main.tf#L155).  But it appears this ENI behavior changed last night on AWS' end.

```
module.ec2_instance.aws_network_interface.primary: Creating...
╷
│ Error: creating EC2 Network Interface: InvalidParameterValue: You can and can only specify one of ipv6-addresses or ipv6-addresses-count  ipv6-prefix and ipv6-prefix-count.
│ 	status code: 400, request id: 7dbba49d-1598-48e0-b88e-901361d2415d
│
│   with module.ec2_instance.aws_network_interface.primary,
│   on module-cml2-ec2-instance/main.tf line 147, in resource "aws_network_interface" "primary":
│  147: resource "aws_network_interface" "primary" {
│
```

I can add the IPv6 address after the fact, but Terraform being declarative, doesn't support this.  So, I'm stuck.  I thought this was limited to bare-metal instances, but no, it's pervasive.  Back away from using IPv6 prefixes, since they will not work in the APIs provided.  I also need more than a `/80` inside Cisco Modeling Labs.  Use a route within the VPC CIDR block and disable source/dest address checking on the ENI.

cc: @rschmied if he's interested.

**Comment by rschmied** [comment]
@cmm-cisco definitely interested, I have an Azure branch ready which I'm going to merge into master soon(ish). I wonder if you have any preferences wrt code separation... e.g. should there be a repository per cloud provider (AWS, Azure, GCP, ...) or should there be an integrated one which let's you choose? The latter is what I currently have adapted for the Azure case. But given your work here I *think* that the integration approach might become more and more complicated.

That said, would you think that the stuff you added in here (is there a "management summary" of changes/capabilities vs. the vanilla repo) is valuable / reusable for others?

**Comment by cmm-cisco** [comment]
Hi @rschmied, I'll definitely be picking up your changes when they're ready.  I want to use CML for security training at a local university that only has TCP port 80/443 connectivity.  I'll have ~20 copies/pods of four VMs, using a bare-metal host on AWS. 

As far as what changes I've made, here's what I have so far:

* Add secrets handling using Cyberark Conjur.  I'm using this for things like CML admin, sysadmin passwords, and CML SmartLicense token.   I think this would be possible to use other secrets managers, but I haven't thought much about how to make this more abstract (like for AWS secrets manager, Hashicorp Vault, etc.).  Conjur's Terraform provider requires you to know all the variable paths ahead of time, which makes it a bit hard to use, imo.  That's why the [config.yml](https://github.com/cmm-cisco/cloud-cml/blob/cmm/config.yml#L28) file looks so odd.
* Add a network load balancer and ACM certificate module (includes Route53 records), to make it easy and nice to access CML.

## Coming next

* I need external v4/v6 connectivity to the Internet for my use case (security training).  IPv6 is somewhat working as an experiment using prefixes, but v4 is not.  I'm [redefining the `default` libvirt network](https://github.com/cmm-cisco/cloud-cml/blob/cmm/module-cml2-ec2-instance/scripts/02-libvirt_default_network.sh#L23) to be routed and disable NAT.  This is a total hack right now.  I'll move NAT functionality into the cloud.
* I plan on moving external routed connectivity to a separate ENI on the CML machine with a separate subnet attached to a egress-only IGW and NAT gateway.
* I'd like to use [lafoe's SSH provider](https://registry.terraform.io/providers/loafoe/ssh/latest/docs) to automate deregistering CML from SmartLicense when Terraform deployment is destroyed, rather than running it by hand.

If you think some of these are useful to others, I'll help however I can.  I'm a bit of a noob when it comes to other clouds like Azure.

**Comment by rschmied** [comment]
I think that secret handling (preferably cloud provider agnostic, Hashi's Vault sounds like a good choice here) and certificate injection is also something I'd like to adopt. Would LetsEncrypt work with AWS and without a specific / own domain? The SSH deregistering idea is also something useful, I'd say.

I am not 100% sure whether I'd want to pick up the IPv6 / networking changes at this point in time. That said, I need to augment the networking in general to also allow cluster support in the future. So maybe that's something to consider. 

But, again, I am still kind of undecided whether an integrated approach makes sense or if I need to separate the different cloud providers due to the differences in implementation (details).

Bottom line: if you have something that you want to contribute back to the "official" tooling then I am more than happy to take a PR!

**Comment by cmm-cisco** [comment]
Closing this in favor of pass-through network loadbalancer on GCP.  It'll do everything we need for BAH.
